include/dynd/kernels/arithmetic.hpp

```diff
@@ -7,42 +7,66 @@
 namespace dynd {
 namespace nd {
 
-  template <type_id_t I0>
-  struct plus_kernel : base_kernel<plus_kernel<I0>, 1> {
+  template <type_id_t I0, typename func_type, func_type f>
+  struct unary_kernel : base_kernel<unary_kernel<I0, func_type, f>, 1> {
     typedef typename type_of<I0>::type A0;
-    typedef decltype(+std::declval<A0>()) R;
-
+    typedef decltype(f(std::declval<A0>())) R;
     DYND_CUDA_HOST_DEVICE void single(char *dst, char *const *src)
     {
-      *reinterpret_cast<R *>(dst) = +*reinterpret_cast<A0 *>(src[0]);
+      *reinterpret_cast<R *>(dst) = f(*reinterpret_cast<A0 *>(src[0]));
     }
   };
 
-  template <type_id_t I0>
-  struct minus_kernel : base_kernel<minus_kernel<I0>, 1> {
-    typedef typename type_of<I0>::type A0;
-    typedef decltype(-std::declval<A0>()) R;
+#define DYND_DeclUnaryOp(OP, NAME)                                                                        \
+  template <typename T>                                                                                   \
+  inline decltype(OP std::declval<T>()) inline_ ## NAME(T val) { return OP val; }                         \
+  template <type_id_t I0>                                                                                 \
+  struct NAME ## _kernel : unary_kernel<I0, decltype(&inline_ ## NAME<typename type_of<I0>::type>),       \
+                                        &inline_ ## NAME<typename type_of<I0>::type>> {};                 \
 
```

Comment:
I prefer the macros primarily because they let the reader know that the *exact* same thing is being done several times over, but that's just me. I can see why we would prefer not to have them though, so we can get rid of them if you'd like.

Thanks for pointing out the kernels there. It makes a lot of sense to use those instead.